File: code_samples/section12/example_8_merge_sort/merge_sort.java
package code_samples.section12.example_8_merge_sort;

import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Stream;

public class merge_sort {

    // void mergeSort(int[] arr) {
    //     int n = arr.length;
    //     int[] tmp = new int[n];
    //     mergeSortRec(arr, tmp, 0, n);
    // }

    /*
        -------------------------------------------------------------------------
        COMMENTED-OUT (baseline) MERGE SORT
        -------------------------------------------------------------------------
        This commented block is the “plain” merge sort implementation without
        any step counting.

        It’s left here on purpose so students can compare:
          - the original algorithm (just sorting),
          - versus the instrumented version below that counts comparisons + copies.

        In a production codebase, you’d typically delete this block to reduce noise.
    */

    // private void mergeSortRec(int[] arr, int[] tmp, int left, int right) {
    //     if (right - left <= 1) return;
    //     int mid = left + (right - left) / 2;
    //     mergeSortRec(arr, tmp, left, mid);
    //     mergeSortRec(arr, tmp, mid, right);
    //     merge(arr, tmp, left, mid, right);
    // }

    // private void merge(int[] arr, int[] tmp, int left, int mid, int right) {
    //     int i = left, j = mid, k = left;
    //     while (i < mid && j < right) {
    //         if (arr[i] <= arr[j]) tmp[k++] = arr[i++];
    //         else                  tmp[k++] = arr[j++];
    //     }
    //     while (i < mid)  tmp[k++] = arr[i++];
    //     while (j < right) tmp[k++] = arr[j++];
    //     for (int p = left; p < right; p++) {
    //         arr[p] = tmp[p];
    //     }
    // }

    // --- step counters for "Big-O" illustration ---
    /*
        These counters instrument the algorithm to help illustrate time complexity.

        compareCount:
          - increments for each element-to-element comparison made during merging:
                arr[i] <= arr[j]

        copyCount:
          - increments for each write into tmp[] during merging
          - increments again for each write copying tmp[] back into arr[]

        Important:
          - These are instance fields, so the counts belong to a specific sorter object.
          - mergeSort() resets them to 0 each time you start a new sort.
    */
    private long compareCount = 0;
    private long copyCount = 0;

    // Public getters so main() can print them
    /*
        Simple accessor methods so callers (like main) can display the final counts.

        This avoids making the fields public while still allowing observation of the
        measured “work” done by the algorithm.
    */
    public long getCompareCount() { return compareCount; }
    public long getCopyCount() { return copyCount; }

    // --- Merge sort API ---
    /*
        mergeSort()
        ----------
        Public entry point:
          - Allocates a temporary buffer tmp[] the same size as arr
          - Resets counters (important for repeat runs)
          - Calls the recursive helper over the full half-open range [0, n)

        Range convention:
          - Uses half-open intervals: [left, right)
            where 'right' is exclusive.
    */
    public void mergeSort(int[] arr) {
        int n = arr.length;
        int[] tmp = new int[n];

        // Reset counters for a clean measurement run
        compareCount = 0;
        copyCount = 0;

        mergeSortRec(arr, tmp, 0, n);
    }

    /*
        mergeSortRec()
        -------------
        Recursively sorts arr[left..right).

        Base case:
          - If the range has 0 or 1 elements (right - left <= 1), it is already sorted.

        Recursive case:
          - Split at mid
          - Sort each half
          - Merge the sorted halves back together
    */
    private void mergeSortRec(int[] arr, int[] tmp, int left, int right) {
        if (right - left <= 1) return;

        int mid = left + (right - left) / 2;

        mergeSortRec(arr, tmp, left, mid);
        mergeSortRec(arr, tmp, mid, right);

        merge(arr, tmp, left, mid, right);
    }

    /*
        merge()
        -------
        Merges two sorted halves into one sorted range.

        Inputs (assumed sorted):
          - arr[left..mid)
          - arr[mid..right)

        Output:
          - arr[left..right) becomes sorted

        Implementation approach:
          1) Merge into tmp[left..right)
          2) Copy tmp[left..right) back into arr[left..right)

        Step counting:
          - compareCount++ for each arr[i] <= arr[j] decision
          - copyCount++ for each write into tmp
          - copyCount++ for each copy back into arr
    */
    private void merge(int[] arr, int[] tmp, int left, int mid, int right) {
        int i = left;   // pointer for left half
        int j = mid;    // pointer for right half
        int k = left;   // pointer for output region in tmp

        // Merge while both halves still have elements
        while (i < mid && j < right) {
            compareCount++; // arr[i] <= arr[j] comparison

            if (arr[i] <= arr[j]) {
                tmp[k++] = arr[i++];
            } else {
                tmp[k++] = arr[j++];
            }

            copyCount++; // wrote one element into tmp
        }

        // Copy remaining elements from left half (if any)
        while (i < mid) {
            tmp[k++] = arr[i++];
            copyCount++;
        }

        // Copy remaining elements from right half (if any)
        while (j < right) {
            tmp[k++] = arr[j++];
            copyCount++;
        }

        // Copy merged block back into original array
        for (int p = left; p < right; p++) {
            arr[p] = tmp[p];
            copyCount++; // count copy back into original array
        }
    }

    // --- File loading helpers ---
    /**
     * Try to load a space-separated integer file from:
     *   1) ..\data\<name> relative to working dir
     *   2) C:\Users\<you>\Desktop\data-structures-algorithms\code_samples\section12\data\<name>
     */
    /*
        loadFile()
        ----------
        Attempts to locate and load an integer file (unordered.txt / ordered.txt).

        Candidate locations:
          1) Relative: ../data/<name>
             - Works when the working directory is the lesson folder.

          2) Explicit path under Desktop:
             - Matches a known local project layout.
             - Uses user.home so it adapts per machine username.

        Behavior:
          - Prints what it’s trying (helpful debugging for path issues).
          - Throws IOException if neither candidate exists.
    */
    private static int[] loadFile(String name) throws IOException {
        // 1) relative path: ..\data\name (works when running from lesson dir)
        Path rel = Paths.get("..", "data", name).normalize();
        System.out.println("Attempting to read (relative): " + rel.toAbsolutePath());

        if (Files.exists(rel)) {
            return readInts(rel);
        }

        // 2) explicit full project path (your known layout)
        Path explicit = Paths.get(
            System.getProperty("user.home"),
            "Desktop",
            "data-structures-algorithms",
            "code_samples",
            "section12",
            "data",
            name
        ).normalize();
        System.out.println("Relative path not found, trying explicit: " + explicit.toAbsolutePath());

        if (Files.exists(explicit)) {
            return readInts(explicit);
        }

        // Neither location exists → fail with a clear error message
        throw new IOException("Missing file '" + name + "' at:\n  "
                + rel.toAbsolutePath() + "\n  " + explicit.toAbsolutePath());
    }

    /*
        readInts()
        ----------
        Reads a text file and parses integers separated by whitespace.

        Implementation notes:
          - Streams file line-by-line with Files.lines() (good for large files).
          - Splits each line on one-or-more whitespace characters (\\s+).
          - Filters out any empty tokens (extra safety).
          - Parses tokens to int and returns a primitive int[].

        try-with-resources:
          - Ensures the Stream (and underlying file handle) is closed automatically.
    */
    private static int[] readInts(Path path) throws IOException {
        try (Stream<String> lines = Files.lines(path)) {
            return lines
                .flatMap(line -> Arrays.stream(line.trim().split("\\s+")))
                .filter(s -> !s.isEmpty())
                .mapToInt(Integer::parseInt)
                .toArray();
        }
    }

    // --- Test harness / demo ---
    /*
        main()
        ------
        Script-style demo:
          1) Print working directory (helps diagnose relative path problems)
          2) Load unordered + ordered reference outputs
          3) Copy unordered into a working array (so original stays intact)
          4) Run merge sort with step counting
          5) Validate correctness by comparing to ordered.txt
          6) Print counters and a short sample output

        Error handling:
          - Any IOException from loadFile/readInts is caught and printed cleanly.
    */
    public static void main(String[] args) {
        System.out.println("Working directory: " + Paths.get("").toAbsolutePath());

        try {
            int[] unordered = loadFile("unordered.txt");
            int[] expected  = loadFile("ordered.txt");

            System.out.println("Loaded unordered length: " + unordered.length);
            System.out.println("Loaded ordered   length: " + expected.length);

            merge_sort sorter = new merge_sort();

            // Work on a copy so the original unordered array remains available if needed
            int[] arr = Arrays.copyOf(unordered, unordered.length);

            sorter.mergeSort(arr);

            // Validate: does our sorted output match the known-good ordered.txt?
            boolean ok = Arrays.equals(arr, expected);
            System.out.println("Sorted correctly? " + ok);

            // Print instrumentation counts
            System.out.println("Comparisons: " + sorter.getCompareCount());
            System.out.println("Copies:      " + sorter.getCopyCount());

            // Show a small sample of the sorted output
            int show = Math.min(10, arr.length);
            System.out.print("First " + show + " sorted values: ");
            for (int i = 0; i < show; i++) {
                System.out.print(arr[i] + (i + 1 < show ? " " : ""));
            }
            System.out.println();

        } catch (IOException e) {
            System.err.println("Error reading input files:");
            System.err.println(e.getMessage());
        }
    }
}

File: code_samples/section12/example_8_merge_sort/merge_sort.js
// function mergeSort(arr) {
//     if (arr.length <= 1) return arr.slice();
//     const mid = Math.floor(arr.length / 2);
//     const left = mergeSort(arr.slice(0, mid));
//     const right = mergeSort(arr.slice(mid));
//     return merge(left, right);
// }

// function merge(left, right) {
//     const result = [];
//     let i = 0, j = 0;
//     while (i < left.length && j < right.length) {
//         if (left[i] <= right[j]) {
//             result.push(left[i++]);
//         } else {
//             result.push(right[j++]);
//         }
//     }
//     return result.concat(left.slice(i)).concat(right.slice(j));
// }

/*
    -------------------------------------------------------------------------
    COMMENTED-OUT (baseline) MERGE SORT
    -------------------------------------------------------------------------
    This block is the original “clean” merge sort that just returns a sorted
    copy of the input.

    It’s commented out so you can compare it to the instrumented version below,
    which adds step counting + file-based testing.

    In real projects you’d remove this, but for learning it’s useful context.
*/

// Example 8 - Merge Sort with Step Counting
import fs from "fs";
import path from "path";
import url from "url";

// ---------------------------------------------------------------------------
// Step counters (global for simplicity)
// ---------------------------------------------------------------------------
/*
    comparisons:
      - increments each time we compare left[i] <= right[j]

    writes:
      - increments each time we write a value into result[]
        (each push counts as one write)

    Note:
      - This version counts writes to the output array only.
      - It does NOT count the cost of slice() allocations, which also copy data.
        (That’s fine for a “Big-O illustration” but worth mentioning.)
*/
let comparisons = 0;
let writes = 0;

/** ------------------------------------------------------------------------
 * merge(left, right)
 * -------------------------------------------------------------------------
 * Merge two sorted arrays into one sorted array.
 *
 * Step counting:
 *  - comparisons++ for each left[i] <= right[j] check
 *  - writes++ for each push into result
 */
function merge(left, right) {
    const result = [];
    let i = 0, j = 0;

    // Merge until one side is exhausted
    while (i < left.length && j < right.length) {
        comparisons++;   // compare left[i] <= right[j]

        if (left[i] <= right[j]) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
        writes++;        // wrote into result[]
    }

    // Copy remaining elements from left
    while (i < left.length) {
        result.push(left[i++]);
        writes++;
    }

    // Copy remaining elements from right
    while (j < right.length) {
        result.push(right[j++]);
        writes++;
    }

    return result;
}

/** ------------------------------------------------------------------------
 * mergeSort(arr)
 * -------------------------------------------------------------------------
 * Recursive merge sort that returns a NEW sorted array (does not sort in-place).
 *
 * Base case:
 *  - length 0 or 1 is already sorted (return a copy with slice()).
 *
 * Recursive case:
 *  - split into halves, sort each, merge the results.
 */
function mergeSort(arr) {
    if (arr.length <= 1) return arr.slice(); // return copy

    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));

    return merge(left, right);
}

// ---------------------------------------------------------------------------
// File Loading Utility (same logic as bubble/selection sort JS examples)
// ---------------------------------------------------------------------------
/*
    loadFile(name)
    --------------
    Robust loader that tries two candidate locations:

      1) ../data/<name> relative to THIS script file
         - works when script lives in lesson folder next to ../data

      2) Explicit Desktop project layout
         - C:\Users\<you>\Desktop\data-structures-algorithms\code_samples\section12\data\<name>

    Why the __dirname stuff?
      - In ES modules, __dirname isn’t available by default, so we build it
        from import.meta.url.
*/
function loadFile(name) {
    const __dirname = path.dirname(url.fileURLToPath(import.meta.url));

    // 1) relative: ../data/<name>
    const rel = path.resolve(__dirname, "..", "data", name);
    console.log("Attempting to read:", rel);

    if (fs.existsSync(rel)) {
        return fs.readFileSync(rel, "utf8")
            .trim()
            .split(/\s+/)
            .map(Number);
    }

    // 2) explicit path (your known layout)
    const explicit = path.resolve(
        process.env.USERPROFILE,
        "Desktop",
        "data-structures-algorithms",
        "code_samples",
        "section12",
        "data",
        name
    );

    console.log("Relative path not found, trying:", explicit);

    if (fs.existsSync(explicit)) {
        return fs.readFileSync(explicit, "utf8")
            .trim()
            .split(/\s+/)
            .map(Number);
    }

    // Neither location worked
    throw new Error(`Missing file '${name}' in:\n  ${rel}\n  ${explicit}`);
}

// ---------------------------------------------------------------------------
// Main Execution / Test Harness
// ---------------------------------------------------------------------------
/*
    main()
    ------
    1) Load unordered.txt and ordered.txt
    2) Reset counters
    3) Sort unordered with merge sort (returns NEW array)
    4) Check correctness vs ordered.txt
    5) Print counters and a small sample output

    Correctness check:
      - Uses JSON stringify to compare arrays.
      - Works fine here because these are flat number arrays.
      - Alternative would be a loop compare (more efficient for large arrays).
*/
function main() {
    try {
        const unordered = loadFile("unordered.txt");
        const expected = loadFile("ordered.txt");

        console.log("Unordered length:", unordered.length);
        console.log("Expected length:", expected.length);

        // reset counters for this run
        comparisons = 0;
        writes = 0;

        const sorted = mergeSort(unordered);

        const correct = JSON.stringify(sorted) === JSON.stringify(expected);
        console.log("Sorted correctly?", correct);

        console.log("Comparisons:", comparisons);
        console.log("Writes:", writes);

        console.log("First 10 sorted elements:", sorted.slice(0, 10).join(" "));
    } catch (err) {
        console.error("Error reading input files:");
        console.error(err.message);
        console.log("Missing input file — aborting.");
    }
}

main();
